server/src/main/java/org/elasticsearch/action/get/TransportGetAction.java

```diff
@@ -298,8 +298,12 @@ private void tryGetFromTranslog(GetRequest request, IndexShard indexShard, Clust
     }
 
     static DiscoveryNode getCurrentNodeOfPrimary(ClusterState clusterState, ShardId shardId) {
-        var shardRoutingTable = clusterState.routingTable().shardRoutingTable(shardId);
-        if (shardRoutingTable.primaryShard() == null || shardRoutingTable.primaryShard().active() == false) {
+        var indexRoutingTable = clusterState.routingTable().index(shardId.getIndex());
+        if (indexRoutingTable == null) {
+            return null;
+        }
+        var shardRoutingTable = indexRoutingTable.shard(shardId.id());
+        if (shardRoutingTable == null || shardRoutingTable.primaryShard() == null || shardRoutingTable.primaryShard().active() == false) {
```

Comment:
For the first case (`shardRoutingTable == null`), I would except a `ShardNotFoundException`, also a 4xx error